realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -0,0 +1,326 @@
+                                          reinterpret_cast<jlong>(new_shared_realm_ptr), j_config_ref.get(),
+                                          j_migration_callback_ref.get(), old_realm->schema_version());
+            };
+        }
+        else {
+            config.migration_function = nullptr;
+        }
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsRealmConfig_nativeSetCompactOnLaunchCallback(
+    JNIEnv* env, jclass, jlong native_ptr, jobject j_compact_on_launch)
+{
+    TR_ENTER_PTR(native_ptr)
+
+    try {
+        auto& config = *reinterpret_cast<Realm::Config*>(native_ptr);
+        if (j_compact_on_launch) {
+            static JavaClass compact_on_launch_class(env, "io/realm/CompactOnLaunchCallback");
+            static JavaMethod should_compact(env, compact_on_launch_class, "shouldCompact", "(JJ)Z");
+            JavaGlobalRef java_compact_on_launch_ref(env, j_compact_on_launch);
+
+            config.should_compact_on_launch_function = [java_compact_on_launch_ref](uint64_t totalBytes,
+                                                                                    uint64_t usedBytes) {
+                JNIEnv* env = JniUtils::get_env(false);
+                return env->CallBooleanMethod(java_compact_on_launch_ref.get(), should_compact,
+                                              static_cast<jlong>(totalBytes), static_cast<jlong>(usedBytes));
+            };
+        }
+        else {
+            config.should_compact_on_launch_function = nullptr;
+        }
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsRealmConfig_nativeSetInitializationCallback(JNIEnv* env,
+                                                                                            jobject j_config,
+                                                                                            jlong native_ptr,
+                                                                                            jobject j_init_callback)
+{
+    TR_ENTER_PTR(native_ptr)
+
+    try {
+        auto& config = *reinterpret_cast<Realm::Config*>(native_ptr);
+
+        if (j_init_callback) {
+            static JavaMethod run_initialization_callback_method(
+                env, get_shared_realm_class(env), "runInitializationCallback",
+                "(JLio/realm/internal/OsRealmConfig;Lio/realm/internal/SharedRealm$InitializationCallback;)V", true);
+            JavaGlobalRef j_init_callback_ref(env, j_init_callback, true);
+            JavaGlobalRef j_config_ref(env, j_config, true);
+            config.initialization_function = [j_init_callback_ref, j_config_ref](SharedRealm realm) {
+                JNIEnv* env = JniUtils::get_env(false);
+                if (env->ExceptionCheck()) {
+                    return;
+                }
+                // Java needs a new pointer for the SharedRealm life control.
+                SharedRealm* new_shared_realm_ptr = new SharedRealm(realm);
```

Comment:
actually in the first implementation, i didn't alloc a new `ShareRealm` ptr here, but just using the pointer to `&realm`. In that way, we gain a very little performance (no need to new a `SharedRealm`), but the code in Java will be a bit confusing. So, do this way to make it looks like other places how we handle the native pointers.